[IO.Mesh.BlenderExporter] Inherit from BaseSimulationExporter

th-skam opened the pull request:
Continuation of PRs to fix: https://github.com/sofa-framework/sofa/issues/5283.

- First two commits change the inheritance of the BlednerExporter from BaseObject to BaseSimulationExporter.
- Last commit is a suggested refactor/clean of BlenderExporter.


______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

th-skam committed:
[BlenderExporter] Example scene added in Component/IO/Mesh

th-skam committed:
[IO.Mesh] Changed inheritance of BlenderExporter to BaseSimulationExporter

th-skam committed:
[BlenderExporter] Refactoring under BaseSimulationExporter

* Substituted member frameCounter with BaseSimulationExporter::m_stepCounter

* The code block that does the writing in BlenderExporter::handleEvent was
  moved to BlenderExporter::write

* fileName is now created through BaseSimulationExporter::getOrCreateTargetPath

* Use of Base::isComponentValid instead of d_componentState

[Scenes] Adjusted BlenderExporter.scn to new variables

th-skam commented:
[ci-build][with-all-tests]

bakpaul commented on Sofa/Component/IO/Mesh/src/sofa/component/io/mesh/BlenderExporter.inl:
```diff
@@ -77,158 +78,155 @@ void BlenderExporter<T>::init()
+    
+        // ############# write data
+    
+        float pos[3]={0,0,0};
```
Why forcing to float instead of using sofa precision (SReal) ?

bakpaul commented on Sofa/Component/IO/Mesh/src/sofa/component/io/mesh/BlenderExporter.inl:
```diff
@@ -77,158 +78,155 @@ void BlenderExporter<T>::init()
+    
+                x0 = x0+x1;
+    
+                pos0[0] = (float)x0[0];
```
You won't need the c style cast if you are using SReal

bakpaul commented on Sofa/Component/IO/Mesh/src/sofa/component/io/mesh/BlenderExporter.inl:
```diff
@@ -77,158 +78,155 @@ void BlenderExporter<T>::init()
+                pos0[1] = (float)x0[1];
+                pos0[2] = (float)x0[2];
+    
+                file.write((char*)pos0,12);
```
Careful this hard coded size of string might break if changing the precision. Maybe in the end, using float might be a constraint of BlenderExporter ?

bakpaul commented on Sofa/Component/IO/Mesh/src/sofa/component/io/mesh/BlenderExporter.inl:
```diff
@@ -77,158 +78,155 @@ void BlenderExporter<T>::init()
+template<class T>
+void BlenderExporter<T>::handleEvent(sofa::core::objectmodel::Event* event)
+{
+    BaseSimulationExporter::handleEvent(event);
```
Don't override it if you call the super class method anyway

th-skam replied:
True that. The thing is, I noticed that other exporters simply override the method and call the parent class, so I assumed this was the convention for the `handleEvent` method. I pushed an additional commit to remove the call in this exporter, as well as in the others (whenever it does nothing but call the parent class method).

th-skam committed:
[IO.Mesh] Removed redundant calls to BaseSimulationExporter::handleEvent

A few of the exporter components in IO/Mesh (BlenderExporter, MeshExporter,
VisualModelOBJExporter) override the handleEvent method yet do nothing except
call the one in the parent class. This call was removed.

th-skam committed:
[BlenderExporter] Implemented intializations in doInit, called by doReInit

th-skam committed:
Update examples/Component/IO/Mesh/BlenderExporter.scn

Co-authored-by: Hugo <hugo.talbot@sofa-framework.org>
